File: crates/yauth/src/repo/password.rs
use super::{RepoFuture, sealed};
use crate::domain;
use uuid::Uuid;

/// Repository for password hashes.
pub trait PasswordRepository: sealed::Sealed + Send + Sync {
    fn find_by_user_id(&self, user_id: Uuid) -> RepoFuture<'_, Option<domain::Password>>;

    /// Insert or replace the password hash for a user.
    fn upsert(&self, input: domain::NewPassword) -> RepoFuture<'_, ()>;
}

/// Repository for email verification tokens.
///
/// # Invariants
///
/// - **`find_by_token_hash`**: MUST return `None` if the token is expired.
pub trait EmailVerificationRepository: sealed::Sealed + Send + Sync {
    fn find_by_token_hash(
        &self,
        token_hash: &str,
    ) -> RepoFuture<'_, Option<domain::EmailVerification>>;

    fn create(&self, input: domain::NewEmailVerification) -> RepoFuture<'_, ()>;

    fn delete(&self, id: Uuid) -> RepoFuture<'_, ()>;

    /// Delete all verifications for a user.
    fn delete_all_for_user(&self, user_id: Uuid) -> RepoFuture<'_, ()>;
}

/// Repository for password reset tokens.
///
/// # Invariants
///
/// - **`find_by_token_hash`**: MUST return `None` if the token is expired or already used.
pub trait PasswordResetRepository: sealed::Sealed + Send + Sync {
    fn find_by_token_hash(
        &self,
        token_hash: &str,
    ) -> RepoFuture<'_, Option<domain::PasswordReset>>;

    fn create(&self, input: domain::NewPasswordReset) -> RepoFuture<'_, ()>;

    /// Delete all unused resets for a user.
    fn delete_unused_for_user(&self, user_id: Uuid) -> RepoFuture<'_, ()>;
}

File: domain/password.go
package domain

import "time"

// Password is a user's stored password hash.
type Password struct {
	UserID       string
	PasswordHash string
}

// NewPassword is the input for upserting a password.
type NewPassword struct {
	UserID       string
	PasswordHash string
}

// EmailVerification is a pending email-verification token.
type EmailVerification struct {
	ID        string
	UserID    string
	TokenHash string
	ExpiresAt time.Time
	CreatedAt time.Time
}

// NewEmailVerification is the input for creating an email verification.
type NewEmailVerification struct {
	ID        string
	UserID    string
	TokenHash string
	ExpiresAt time.Time
	CreatedAt time.Time
}

// PasswordReset is a pending password-reset token.
type PasswordReset struct {
	ID        string
	UserID    string
	TokenHash string
	ExpiresAt time.Time
	UsedAt    *time.Time
	CreatedAt time.Time
}

// NewPasswordReset is the input for creating a password reset.
type NewPasswordReset struct {
	ID        string
	UserID    string
	TokenHash string
	ExpiresAt time.Time
	CreatedAt time.Time
}
